File: src/algorithms/treeandgraph/PathSum.java
package algorithms.treeandgraph; 

import java.util.*; 
import lib.TreeNode; 

public class PathSum {
    /* Function: Find if there exist a path from root to a leaf that sum up to a target. 
     * Solution: Notice root cannot count as leaf. 
     */ 
    
    public boolean hasPathSum(TreeNode root, int sum) {
        if (root == null) {
            return false; 
        }
        
        return pathHelper(root, sum); 
    }
    
    private boolean pathHelper(TreeNode root, int sum) {
        if (root.left == null && root.right == null) {
            return sum == root.val; 
        }
        
        sum -= root.val; 
        boolean res = false; 
        if (root.left != null) {
            res |= pathHelper(root.left, sum); 
        }
        if (root.right != null) {
            res |= pathHelper(root.right, sum); 
        }
        
        return res; 
    }
} 

File: src/CPP/src/leetcode/PathSum.cpp
/*
 * [Source] https://leetcode.com/problems/path-sum/
 * [Difficulty]: Easy
 * [Tag]: Tree
 * [Tag]: Depth-first Search
 */

#include <iostream>

using namespace std;

// [Solution]: DFS. backtracking.
// [Corner Case]:
class Solution {
};

/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/treeandgraph/PathSum.java
 */

/* Java solution
public class Solution {
    public boolean hasPathSum(TreeNode root, int sum) {
		if(root==null) return false;
		return pathSum(root, sum);
    }
	
	public boolean pathSum(TreeNode root, int sum){
		if(root==null) return false;
		if((sum-root.val)==0 && root.left==null && root.right==null) return true; // should be leaf
		boolean l=pathSum(root.left, sum-root.val);
		boolean r=pathSum(root.right, sum-root.val);
		return l || r;
    }
}
 */

int main() {
    Solution sol;

    return 0;
}
